File: chat-account/src/main/java/com/devanmejia/chataccount/config/security/JWTAuthenticationManager.java
package com.devanmejia.chataccount.config.security;

import com.devanmejia.chataccount.exception.AuthException;
import com.devanmejia.chataccount.config.security.user_details.UserDetailsService;
import com.devanmejia.chataccount.service.user.UserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.userdetails.UserDetails;

@Configuration
public class JWTAuthenticationManager implements AuthenticationManager {
    private final UserDetailsService userDetailsService;
    private final UserService userService;

    @Autowired
    public JWTAuthenticationManager(UserDetailsService userDetailsService, UserService userService) {
        this.userDetailsService = userDetailsService;
        this.userService = userService;
    }

    @Override
    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
        String token = (String) authentication.getCredentials();
        try{
            String login = userDetailsService.getUserLogin(token);
            UserDetails userDetails = userService.findByLogin(login);
            return new UsernamePasswordAuthenticationToken(userDetails,
                    "", userDetails.getAuthorities());
        } catch (AuthException e){
            return new UsernamePasswordAuthenticationToken(null, null);
        }
    }
}

File: chat-messaging/src/main/kotlin/com/devanmejia/chatmessaging/configuration/security/JWTAuthenticationManager.kt
package com.devanmejia.chatmessaging.configuration.security

import com.devanmejia.chatmessaging.exception.AuthException
import com.devanmejia.chatmessaging.service.UserDetailsService
import kotlinx.coroutines.reactor.mono
import org.springframework.security.authentication.ReactiveAuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono

@Component
class JWTAuthenticationManager(private val userDetailsService: UserDetailsService)
    : ReactiveAuthenticationManager {

    override fun authenticate(authentication: Authentication): Mono<Authentication> {
        return mono {
            val token = authentication.credentials as String
            try {
                val userDetails = userDetailsService.getUserDetails(token)
                UsernamePasswordAuthenticationToken(userDetails, "", userDetails.authorities)
            } catch (ex: AuthException){
                UsernamePasswordAuthenticationToken(null, null)
            }
        }
    }
}

